File: src/main/java/qmech/lib/objects/meta/MetaItemBase.java
package qmech.lib.objects.meta;

import com.google.common.base.Preconditions;
import com.google.common.collect.Maps;
import cpw.mods.fml.common.registry.GameRegistry;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.client.renderer.texture.IIconRegister;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.IIcon;
import net.minecraft.world.World;
import qmech.mod.Reference;

import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

public abstract class MetaItemBase extends Item {
    public final String prefix;
    private final Map<Integer, IMetaItemBase> metaitems = Maps.newHashMap();
    protected MetaItemBase(String intName) {
        this.setHasSubtypes(true);
        this.setMaxDamage(0);
        this.setUnlocalizedName(intName);
        this.prefix = intName;
        GameRegistry.registerItem(this, intName);
    }

    public void registerItem(int id, IMetaItemBase item) {
        IMetaItemBase prev = this.metaitems.put(id, item);
        Preconditions.checkState(prev == null, "Config error: replacing meta item %s with %s", prev, item);
    }

    @Override
    public void registerIcons(IIconRegister register) {
        for (IMetaItemBase item : this.metaitems.values()) {
            item.registerIcons(register, this.prefix);
        }
    }

    @Override
    public IIcon getIconFromDamage(int i) {
        IMetaItemBase meta = this.getMeta(i);
        if (meta != null) {
            return meta.getIcon();
        }
        return null;
    }

    @Override
    public String getUnlocalizedName(ItemStack stack) {
        IMetaItemBase meta = this.getMeta(stack.getItemDamage());
        if (meta != null) {
            return String.format("item.%s_%s", this.prefix, meta.getUnlocalizedName(stack));
        }
        return "";
    }

    @Override
    public boolean onItemUse(ItemStack itemStack, EntityPlayer player, World world, int x, int y, int z, int side, float par8, float par9, float par10) {
        IMetaItemBase meta = this.getMeta(itemStack.getItemDamage());
        return meta == null || meta.onItemUse(itemStack, player, world, x, y, z, side, par8, par9, par10);
    }

    @Override
    public ItemStack onItemRightClick(ItemStack itemStack, World world, EntityPlayer player) {
        IMetaItemBase meta = this.getMeta(itemStack.getItemDamage());
        if (meta != null) {
            return meta.onItemRightClick(itemStack, player, world);
        }
        return itemStack;
    }

    @Override
    public boolean hitEntity(ItemStack itemStack, EntityLivingBase target, EntityLivingBase player) {
        IMetaItemBase meta = this.getMeta(itemStack.getItemDamage());
        return meta == null || meta.hitEntity(itemStack, target, player);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public boolean hasEffect(ItemStack itemStack, int pass) {
        IMetaItemBase meta = this.getMeta(itemStack.getItemDamage());
        return (meta != null) && meta.hasEffect(pass);
    }

    @Override
    @SideOnly(Side.CLIENT)
    @SuppressWarnings({"rawtypes", "unchecked"})
    public void getSubItems(Item item, CreativeTabs tab, List subItems) {
        for (Entry<Integer, IMetaItemBase> entry : this.metaitems.entrySet())
            entry.getValue().addToCreativeList(item, entry.getKey(), subItems);
    }

    IMetaItemBase getMeta(int id) {
        return this.metaitems.get(id);
    }

    public IMetaItemBase getMeta(ItemStack itemStack) {
        return this.getMeta(itemStack.getItemDamage());
    }

    protected ItemStack newItemStack(int id) {
        return this.newItemStack(id, 1);
    }

    ItemStack newItemStack(int id, int number) {
        return new ItemStack(this, number, id);
    }

    public ItemStack newItemStack(IMetaItemBase meta, int size) {
        for (Entry<Integer, IMetaItemBase> o : this.metaitems.entrySet()) {
            if (o.getValue().equals(meta)) {
                return this.newItemStack(o.getKey(), size);
            }
        }
        return null;
    }

    public static class MetaItem implements IMetaItemBase {

        private final String mod;
        private final String name;
        private IIcon icon;
        private final Object[] recipes;
        private boolean visibleInCreative = true;

        public MetaItem(String name, Object... recipes) {
            this.mod = Reference.MOD_ID;
            this.name = name;
            this.recipes = recipes;
        }

        public MetaItem hideFromCreative() {
            this.visibleInCreative = false;
            return this;
        }

        @Override
        public IIcon getIcon() {
            return this.icon;
        }

        @Override
        public String getUnlocalizedName(ItemStack stack) {
            return this.name;
        }

        @Override
        public boolean hitEntity(ItemStack itemStack, EntityLivingBase target, EntityLivingBase player) {
            return false;
        }

        @Override
        public boolean onItemUse(ItemStack itemStack, EntityPlayer player, World world, int x, int y, int z, int side, float par8, float par9, float par10) {
            return false;
        }

        @Override
        public ItemStack onItemRightClick(ItemStack itemStack, EntityPlayer player, World world) {
            return itemStack;
        }

        @Override
        public void registerIcons(IIconRegister register, String prefix) {
            this.registerIcon(register, String.format("%s_%s", prefix, this.name));
        }

        void registerIcon(IIconRegister register, String name) {
            this.icon = register.registerIcon(String.format("%s:%s", this.mod, name));
        }

        @Override
        public void addToCreativeList(Item item, int meta, List<ItemStack> result) {
            if (this.visibleInCreative) {
                result.add(new ItemStack(item, 1, meta));
            }
        }

        @Override
        public boolean hasEffect(int renderPass) {
            return false;
        }

    }
}

File: src/main/scala/geek/lawsof/physics/lib/item/meta/MetaItemBase.scala
package geek.lawsof.physics.lib.item.meta

import net.minecraft.item.{ItemStack, Item}
import cpw.mods.fml.common.registry.GameRegistry
import net.minecraft.util.IIcon
import cpw.mods.fml.relauncher.{SideOnly, Side}
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.world.World
import net.minecraft.client.renderer.texture.IIconRegister
import scala.collection.mutable
import java.util

/**
 * Created by anshuman on 26-05-2014.
 */
abstract class MetaItemBase(intName: String) extends Item {
  setHasSubtypes(true)
  setMaxDamage(0)
  this.setUnlocalizedName(intName)
  GameRegistry.registerItem(this, intName)

  def registerItem(id: Int, item: IMetaSubItem) = metaitems.put(id, item)

  override def registerIcons(register: IIconRegister) {
    for (item <- metaitems.values) {
      item.registerIcons(register, intName)
    }
  }

  override def getIconFromDamage(i: Int): IIcon = {
    if (metaitems contains i) metaitems(i).getIcon else metaitems(0).getIcon
  }

  override def getUnlocalizedName(stack: ItemStack): String = {
    if (metaitems contains stack.getItemDamage) s"item.${intName}_${metaitems(stack.getItemDamage).getUnlocalizedName(stack)}" else "Invalid Item"
  }

  override def getItemStackDisplayName(par1ItemStack: ItemStack): String = if (getUnlocalizedName(par1ItemStack).equals("Invalid Item")) "Invalid Item" else super.getItemStackDisplayName(par1ItemStack)

  override def onItemUse(itemStack: ItemStack, player: EntityPlayer, world: World, x: Int, y: Int, z: Int, side: Int, par8: Float, par9: Float, par10: Float): Boolean = {
    val meta: IMetaSubItem = getMeta(itemStack.getItemDamage)
    if (meta != null) {
      return meta.onItemUse(itemStack, player, world, x, y, z, side, par8, par9, par10)
    }
    true
  }

  override def onItemRightClick(itemStack: ItemStack, world: World, player: EntityPlayer): ItemStack = {
    val meta: IMetaSubItem = getMeta(itemStack.getItemDamage)
    if (meta != null) {
      return meta.onItemRightClick(itemStack, player, world)
    }
    itemStack
  }

  override def hitEntity(itemStack: ItemStack, target: EntityLivingBase, player: EntityLivingBase): Boolean = {
    val meta: IMetaSubItem = getMeta(itemStack.getItemDamage)
    if (meta != null) {
      return meta.hitEntity(itemStack, target, player)
    }
    true
  }

  @SideOnly(Side.CLIENT) override def hasEffect(itemStack: ItemStack, pass: Int): Boolean = {
    if (metaitems contains itemStack.getItemDamage) metaitems(itemStack.getItemDamage).hasEffect(pass) else false
  }

  @SideOnly(Side.CLIENT)
  override def getSubItems(item: Item, tab: CreativeTabs, subItems: util.List[_]) {
    import scala.collection.JavaConversions._
    for (entry <- metaitems.entrySet) entry.getValue.addToCreativeList(item, entry.getKey, subItems.asInstanceOf[util.List[ItemStack]])
  }

  def getMeta(i: Int): IMetaSubItem = {
    if (metaitems.get(i).nonEmpty) metaitems.get(i).get else metaitems(0)
  }

  def getMeta(itemStack: ItemStack): IMetaSubItem = {
    getMeta(itemStack.getItemDamage)
  }

  def newItemStack(id: Int): ItemStack = {
    newItemStack(id, 1)
  }

  def newItemStack(id: Int, number: Int): ItemStack = {
    new ItemStack(this, number, id)
  }

  def newItemStack(meta: IMetaSubItem, size: Int): ItemStack = {
    import scala.collection.JavaConversions._
    for (o <- metaitems.entrySet) {
      if (o.getValue == meta) {
        return newItemStack(o.getKey, size)
      }
    }
    null
  }

  protected var metaitems: mutable.HashMap[Integer, IMetaSubItem] = new mutable.HashMap[Integer, IMetaSubItem]
  var prefix: String = null
}
